Tribler/community/tunnel/hidden_community.py

```diff
@@ -186,9 +193,32 @@ def monitor_downloads(self, dslist):
             old_state = self.download_states.get(info_hash, None)
             state_changed = new_state != old_state
 
-            force_dht_lookup = (time.time() - self.last_dht_lookup.get(info_hash, 0)) >= 300
+            time_elapsed = (time.time() - self.last_dht_lookup.get(info_hash, 0))
+            force_dht_lookup = time_elapsed >= self.settings.dht_lookup_interval
+
+            # If the introducing circuit does not exist anymore or timed out: Built a new circuit
+            if info_hash in self.seeding_introduction:
+                circuit_id, required_exit = self.seeding_introduction[info_hash]
+                if circuit_id not in self.my_intro_points and info_hash not in self.infohashes_ip_waiting_for_circuit:
+                    self._logger.debug('Recreate the introducing circuit for %s' % info_hash.encode('hex'))
+                    # check whether required_exit is still a candidate here
+                    if (required_exit[0], required_exit[1]) not in self._candidates:
+                        self._logger.debug('Candidate for exiting introducing circuit does not exist anymore')
+                        self.seeding_introduction.pop(info_hash)
+                        required_exit = None
```

Comment:
I'll look into that - but problem with pinging them is that it should ping all candidates, no matter if they have circuits or not. Otherwise it leaks information about the initiator of a circuit to the exit node?